### PR TITLE
feat: Implement single-flow account recovery per spec

### PR DESCRIPTION
**Pull request by timwonderer**
## Schema Change Gate Classification (required for schema changes)

- [x] **EXPAND** – Additive, backward-compatible (no removals)

## Description

Revises the account recovery feature to follow the streamlined single-flow spec in `docs/development/account_recovery_and_reclaim_spec.md`. The previous implementation had two separate flows (Flow 1: forgot username, Flow 2: forgot PIN/passphrase). This PR consolidates into a single teacher-initiated, time-bounded recovery flow:

1. **Teacher initiates reset** — generates 8-char hex code with 10-minute TTL, sets `recovery_status = 'to_be_claimed'`
2. **Student enters join_code + reset_code** — validated against expiration, status, and join_code scoping
3. **Student re-registers identity** — PII updated, salt rotated, credentials cleared for re-setup
4. **Student sets new credentials** — via existing setup flow (username → PIN → passphrase)
5. **Completion** — recovery state cleared, `recovery_status = 'active'`

### Key changes:
- **`app/routes/recovery.py`** — Complete rewrite: removed Flow 2 routes, added `to_be_claimed` status validation, fixed session key bug (`claimed_student_id` instead of `student_id`), syncs PII to TeacherBlock entries
- **`app/routes/admin.py`** — Admin edit modal now generates recovery codes instead of directly clearing credentials; fixed `onboarding_status()` early return for admins with no class periods
- **`templates/student_detail.html`** — Added "Account Recovery In Progress" section showing reset code and join code when recovery is active
- **`templates/student/recovery/`** — Simplified landing page to single flow, updated labels
- **`templates/student_login.html`** — Changed recovery link text to "I can't log into my account"
- **`tests/conftest.py`** — Fixed `client_with_fk` fixture to properly enable SQLite FK enforcement via PRAGMA
- **`tests/test_announcements.py`** — Fixed cascade delete test to use `client_with_fk` and query-based assertion

## Type of Change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Refactoring (no functional changes)
- [ ] Performance improvement
- [ ] Other (please describe):

## Testing

- [x] Tested locally
- [x] All existing tests pass
- [x] Added new tests for new functionality

**Test results:** 424 passed, 0 failed, 1 skipped, 2 warnings (pre-existing)

### New test coverage (17 tests in `test_student_recovery.py` + 1 E2E in `test_flow_credential_reset.py`):

**Step 1 — Teacher initiates reset:**
- `test_teacher_generates_reset_code` — verifies code generation, TTL, status transition
- `test_multiple_resets_invalidate_prior_codes` — verifies old codes are replaced

**Step 2 — Student lookup:**
- `test_student_lookup_success` — valid join_code + reset_code → redirect
- `test_student_lookup_wrong_join_code` — wrong join_code → generic error
- `test_student_lookup_expired_code` — expired code → generic error
- `test_student_lookup_wrong_status` — status != to_be_claimed → generic error
- `test_student_lookup_nonexistent_code` — invalid code → generic error

**Step 3 — Identity re-registration:**
- `test_verify_identity_updates_pii_and_clears_credentials` — PII updated, salt rotated, credentials cleared
- `test_verify_identity_no_new_student_row` — no new student record created
- `test_verify_identity_redirects_without_session` — no session → redirect
- `test_verify_identity_missing_fields` — validation enforced

**Economic invariance:**
- `test_recovery_preserves_balance_and_transactions` — balance/transactions unchanged

**Security:**
- `test_reset_code_invalid_after_use` — single-use enforcement
- `test_only_one_active_reset_code_per_student` — one code at a time

**Edge cases:**
- `test_reclaim_when_not_to_be_claimed_fails` — status guard
- `test_interrupting_reclaim_does_not_corrupt_record` — mid-flow abandonment safety

**End-to-end:**
- `test_credential_reset_flow` — full 5-step flow with economic data verification

### Bug fixes verified:
- Fixed `client_with_fk` fixture (4 cascade delete tests now pass)
- Fixed `onboarding_status()` early return (1 test now passes)

## Database Migration Checklist

**Does this PR include a database migration?** [x] Yes / [ ] No

If **Yes**, confirm:

- [x] Synced with `main` branch immediately before running `flask db migrate`
- [x] Migration file reviewed and verified correct `down_revision`
- [x] Tested `flask db upgrade` successfully
- [x] Tested `flask db downgrade` successfully
- [x] Confirmed only ONE migration head exists (pre-push hook should verify this)
- [x] Migration has a descriptive message/filename
- [x] Breaking changes or data migrations documented in PR description

**Migration file location:**
`migrations/versions/d27f7eb9d45_add_account_recovery_fields_to_student.py`

## Checklist

- [x] My code follows the project's style guidelines
- [x] I have performed a self-review of my own code
- [x] I have commented my code where necessary, particularly in hard-to-understand areas
- [x] I have updated the documentation accordingly
- [x] My changes generate no new warnings or errors
- [x] I have read and followed the [contributing guidelines](../CONTRIBUTING.md)

## Related Issues

Implements the account recovery & reclaim spec defined in `docs/development/account_recovery_and_reclaim_spec.md`.

## Additional Notes

- The 2 warnings in test output are pre-existing (`SADeprecationWarning` for `DISTINCT ON` in SQLite and a `DeprecationWarning` for `datetime.utcnow()`). Neither was introduced by this PR.
- The `client_with_fk` fixture fix and `onboarding_status()` fix are included because they resolved pre-existing test failures discovered during the regression testing for this feature.

### CODE REVIEW

File: scripts/create_admin.py [hunk 1]
```diff
@@ -40,12 +63,30 @@
         print()
         print("TOTP Setup Instructions:")
         print("1. Open your authenticator app (Google Authenticator, Authy, etc.)")
-        print("2. Scan the QR code or manually enter the secret key")
+        print("2. Scan the QR code below or manually enter the secret key:")
+        print()
+        print(f"   SECRET KEY: {totp_secret}")
```
[the remark this answers is not in the document]
Comment (timwonderer) [reply scripts/create_admin.py]: Addressed in commit 9e39929. This thread has been fixed in the latest push (security/logging, recovery hardening, JS scope fix, query optimization, and cleanup updates as applicable to this file).

File: templates/student_detail.html [hunk 2]
```diff
@@ -868,31 +990,32 @@ <h6 class="mb-0 text-warning"><i class="bi bi-key me-2"></i>Login Management</h6
...
+        } else {
+            showToast(data.error || 'Failed to load tap entries', true);
         }
-    });
+    } catch (error) {
+        showToast('Network error', true);
+    } finally {
+        btn.disabled = false;
+        btn.innerHTML = '<i class="bi bi-arrow-clockwise me-1"></i>Reload Tap Entries';
+    }
```
[the remark this answers is not in the document]
Comment (timwonderer) [reply templates/student_detail.html]: Addressed in commit 9e39929. This thread has been fixed in the latest push (security/logging, recovery hardening, JS scope fix, query optimization, and cleanup updates as applicable to this file).

File: app/routes/recovery.py [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+        TeacherBlock.query.filter_by(student_id=student.id).update({
+            'first_name': student.first_name,
+            'last_initial': student.last_initial,
+            'last_name_hash_by_part': student.last_name_hash_by_part or [],
+            'dob_sum': student.dob_sum or 0,
+            'salt': new_salt,
+            'first_half_hash': student.first_half_hash,
+            'is_claimed': False,
+            'claimed_at': None,
+        })
```
[the remark this answers is not in the document]
Comment (timwonderer) [reply app/routes/recovery.py]: Addressed in commit 9e39929. This thread has been fixed in the latest push (security/logging, recovery hardening, JS scope fix, query optimization, and cleanup updates as applicable to this file).

File: app/routes/recovery.py [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+
+        if not valid:
+            flash("Invalid or expired recovery code.", "error")
+            return redirect(url_for('recovery.account_lookup'))
+
+        # Success: store student_id in session for next step
+        session['recovery_student_id'] = student.id
+        return redirect(url_for('recovery.verify_identity'))
+
+    return render_template('student/recovery/account_lookup.html')
```
[the remark this answers is not in the document]
Comment (timwonderer) [reply app/routes/recovery.py]: Addressed in commit 9e39929. This thread has been fixed in the latest push (security/logging, recovery hardening, JS scope fix, query optimization, and cleanup updates as applicable to this file).

File: scripts/create_admin.py [hunk 5]
```diff
@@ -6,14 +6,37 @@
...
 import qrcode
+import select
+import time
 from io import BytesIO
 
 # Set up Flask app context
 from app import app
 from app.extensions import db
 from app.models import SystemAdmin, Admin
+from app.models import SystemAdmin, Admin
```
[the remark this answers is not in the document]
Comment (timwonderer) [reply scripts/create_admin.py]: Addressed in commit 9e39929. This thread has been fixed in the latest push (security/logging, recovery hardening, JS scope fix, query optimization, and cleanup updates as applicable to this file).

File: scripts/create_admin.py [hunk 6]
```diff
@@ -40,12 +63,30 @@ def create_system_admin(username):
...
+        print()
+        print(f"   SECRET KEY: {totp_secret}")
+        print()
+        
+        # Generate and print QR code
+        try:
+            uri = pyotp.start(totp_secret, user=username, issuer_name="Classroom Economy - System", provisioning_uris=True)
+            # PyOTP start returns provisioning URI if provisioning_uris=True? 
+            # No, pyotp.totp.TOTP... let's use standard pyotp usage or just manually build it if needed.
+            # Actually pyotp.totp.TOTP(totp_secret).provisioning_uri(...) is standard.
```
[the remark this answers is not in the document]
Comment (timwonderer) [reply scripts/create_admin.py]: Addressed in commit 9e39929. This thread has been fixed in the latest push (security/logging, recovery hardening, JS scope fix, query optimization, and cleanup updates as applicable to this file).

File: templates/student/recovery/credential_lookup.html [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    <div class="mb-4">
+        <label class="form-label text-secondary small fw-bold">Teacher Reset Code</label>
+        <input type="text" name="reset_code" class="form-control form-control-lg text-uppercase"
+            placeholder="8-character code" required autocomplete="off">
+        <div class="form-text">Ask your teacher for this code. It expires in 10 minutes.</div>
+    </div>
+
+    <button type="submit" class="btn btn-primary w-100 btn-lg">Verify</button>
+</form>
+{% endblock %}
```
[the remark this answers is not in the document]
Comment (timwonderer) [reply templates/student/recovery/credential_lookup.html]: Addressed in commit 9e39929. This thread has been fixed in the latest push (security/logging, recovery hardening, JS scope fix, query optimization, and cleanup updates as applicable to this file).

File: templates/student_detail.html [hunk 8]
```diff
@@ -868,31 +990,32 @@ <h6 class="mb-0 text-warning"><i class="bi bi-key me-2"></i>Login Management</h6
...
 
     // Load tap entries button
-    document.getElementById('loadTapEntries')?.addEventListener('click', async function() {
+    document.getElementById('loadTapEntries')?.addEventListener('click', async function () {
         const studentId = {{ student.id }};
-        const container = document.getElementById('tapEntriesContainer');
-        const btn = this;
+    };
+    const container = document.getElementById('tapEntriesContainer');
+    const btn = this;
```
[the remark this answers is not in the document]
Comment (timwonderer) [reply templates/student_detail.html]: Addressed in commit 9e39929. This thread has been fixed in the latest push (security/logging, recovery hardening, JS scope fix, query optimization, and cleanup updates as applicable to this file).

File: scripts/create_admin.py [hunk 9]
```diff
@@ -40,12 +63,30 @@ def create_system_admin(username):
...
+            
+            uri = pyotp.totp.TOTP(totp_secret).provisioning_uri(name=username, issuer_name="Classroom Economy - System")
```
[the remark this answers is not in the document]
Comment (timwonderer) [reply scripts/create_admin.py]: Addressed in commit 9e39929. This thread has been fixed in the latest push (security/logging, recovery hardening, JS scope fix, query optimization, and cleanup updates as applicable to this file).

File: app/routes/recovery.py [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+    student_id remains unchanged. No new student record is created.
+    Operation is atomic.
+
+    Then proceeds to Step 4 — Credential Re-Establishment
+    (normal credential setup: username, PIN, passphrase).
+    """
+    student_id = session.get('recovery_student_id')
+    if not student_id:
+        return redirect(url_for('recovery.account_lookup'))
+
```
[the remark this answers is not in the document]
Comment (timwonderer) [reply app/routes/recovery.py]: Addressed in commit de12118. Implemented the requested change for this thread (recovery step revalidation, state-machine/migration alignment, lookup hardening, fixture/test cleanup, and import/lint fixes as applicable).

File: app/routes/recovery.py [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+        student.username_lookup_hash = None
+        student.pin_hash = None
+        student.passphrase_hash = None
+        student.has_completed_setup = False
+
+        # Step 5 — Completion: clear recovery state, set status -> active
+        student.reset_code = None
+        student.reset_code_expires_at = None
+        student.recovery_status = 'active'
+
```
[the remark this answers is not in the document]
Comment (timwonderer) [reply app/routes/recovery.py]: Addressed in commit de12118. Implemented the requested change for this thread (recovery step revalidation, state-machine/migration alignment, lookup hardening, fixture/test cleanup, and import/lint fixes as applicable).

File: migrations/versions/f5bf397b9d45_add_account_recovery_fields_to_student.py [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+depends_on = None
+
+
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    with op.batch_alter_table('students', schema=None) as batch_op:
+        batch_op.add_column(sa.Column('reset_code', sa.String(length=8), nullable=True))
+        batch_op.add_column(sa.Column('reset_code_expires_at', sa.DateTime(timezone=True), nullable=True))
+        batch_op.add_column(sa.Column('money_action_cooldown_until', sa.DateTime(timezone=True), nullable=True))
+        batch_op.add_column(sa.Column('recovery_status', sa.String(length=20), server_default='none', nullable=False))
```
[the remark this answers is not in the document]
Comment (timwonderer) [reply migrations/versions/f5bf397b9d45_add_account_recovery_fields_to_student.py]: Addressed in commit de12118. Implemented the requested change for this thread (recovery step revalidation, state-machine/migration alignment, lookup hardening, fixture/test cleanup, and import/lint fixes as applicable).

File: app/routes/recovery.py [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+    """
+    # Verify scoping: admin must have access to this student
+    student = get_student_for_admin(student_id)
+    if not student:
+        flash("Student not found.", "error")
+        return redirect(url_for('admin.students'))
+
+    # Invalidate any existing reset_code, then generate new one
+    code = secrets.token_hex(4).upper()  # 8-char hex string
+
```
[the remark this answers is not in the document]
Comment (timwonderer) [reply app/routes/recovery.py]: Addressed in commit de12118. Implemented the requested change for this thread (recovery step revalidation, state-machine/migration alignment, lookup hardening, fixture/test cleanup, and import/lint fixes as applicable).

File: app/routes/recovery.py [hunk 14]
```diff
@@ -0,0 +1,208 @@
...
+        reset_code = request.form.get('reset_code', '').strip().upper()
+
+        if not join_code or not reset_code:
+            flash("Both fields are required.", "error")
+            return redirect(url_for('recovery.account_lookup'))
+
+        # Find student by reset_code
+        student = Student.query.filter_by(reset_code=reset_code).first()
+
+        # Validate all conditions — use a single generic error for security
```
[the remark this answers is not in the document]
Comment (timwonderer) [reply app/routes/recovery.py]: Addressed in commit de12118. Implemented the requested change for this thread (recovery step revalidation, state-machine/migration alignment, lookup hardening, fixture/test cleanup, and import/lint fixes as applicable).

File: tests/conftest.py [hunk 15]
```diff
@@ -29,70 +29,72 @@
...
-    # Foreign key constraints are enabled by the event listener.
-    
-    client = flask_app.test_client()
+    from sqlalchemy import event
+
+    dialect = db.engine.dialect.name
+    if dialect == 'sqlite':
+        @event.listens_for(db.engine, "connect")
+        def _set_sqlite_pragma(dbapi_connection, connection_record):
+            cursor = dbapi_connection.cursor()
```
[the remark this answers is not in the document]
Comment (timwonderer) [reply tests/conftest.py]: Addressed in commit de12118. Implemented the requested change for this thread (recovery step revalidation, state-machine/migration alignment, lookup hardening, fixture/test cleanup, and import/lint fixes as applicable).

File: scripts/create_admin.py [hunk 16]
```diff
@@ -6,14 +6,37 @@
 import sys
 import pyotp
 import qrcode
+import select
+import time
```
[the remark this answers is not in the document]
Comment (timwonderer) [reply scripts/create_admin.py]: Addressed in commit de12118. Implemented the requested change for this thread (recovery step revalidation, state-machine/migration alignment, lookup hardening, fixture/test cleanup, and import/lint fixes as applicable).

File: tests/test_flow_credential_reset.py [hunk 17]
```diff
@@ -0,0 +1,193 @@
...
+  6. Economic data is preserved
+"""
+import pytest
+from werkzeug.security import generate_password_hash
+from app import db
+from app.models import Admin, Student, TeacherBlock, StudentBlock, Transaction
+from app.hash_utils import hash_username, hash_username_lookup, get_random_salt
+from app.utils.name_utils import hash_last_name_parts
+from app.utils.claim_credentials import compute_primary_claim_hash
+from app.utils.time import utc_now
```
[the remark this answers is not in the document]
Comment (timwonderer) [reply tests/test_flow_credential_reset.py]: Addressed in commit de12118. Implemented the requested change for this thread (recovery step revalidation, state-machine/migration alignment, lookup hardening, fixture/test cleanup, and import/lint fixes as applicable).

File: tests/test_student_recovery.py [hunk 18]
```diff
@@ -0,0 +1,487 @@
...
+  - Teacher-initiated reset code generation (Step 1)
+  - Student lookup via join_code + reset_code (Step 2)
+  - Identity re-registration (Step 3)
+  - Credential re-establishment (Step 4 — via existing setup flow)
+  - Completion (Step 5)
+  - State machine transitions
+  - Security & edge cases from the spec acceptance tests
+"""
+import pytest
+from datetime import timedelta, timezone
```
[the remark this answers is not in the document]
Comment (timwonderer) [reply tests/test_student_recovery.py]: Addressed in commit de12118. Implemented the requested change for this thread (recovery step revalidation, state-machine/migration alignment, lookup hardening, fixture/test cleanup, and import/lint fixes as applicable).

File: tests/verify_onboarding_fix.py [hunk 19]
```diff
@@ -0,0 +1,78 @@
+
+import os
+import pytest
+from app import create_app, db
+from app.models import Admin, TeacherOnboarding, TeacherBlock
+from flask import session
```
[the remark this answers is not in the document]
Comment (timwonderer) [reply tests/verify_onboarding_fix.py]: Addressed in commit de12118. Implemented the requested change for this thread (recovery step revalidation, state-machine/migration alignment, lookup hardening, fixture/test cleanup, and import/lint fixes as applicable).

File: tests/conftest.py [hunk 20]
```diff
@@ -29,70 +29,72 @@
...
         ENV="testing",
         SESSION_COOKIE_SECURE=False,
     )
+    
+    # Ensure strict separation - if connection fails, test fails
+    with flask_app.app_context():
+        
+        db.drop_all() 
+        db.create_all()
+        pass
```
[the remark this answers is not in the document]
Comment (timwonderer) [reply tests/conftest.py]: Addressed in commit de12118. Implemented the requested change for this thread (recovery step revalidation, state-machine/migration alignment, lookup hardening, fixture/test cleanup, and import/lint fixes as applicable).
